keras_nlp/models/roberta.py

```diff
@@ -0,0 +1,298 @@
+# Copyright 2022 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Roberta model configurable class, preconfigured versions, and task heads."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.layers import TokenAndPositionEmbedding
+from keras_nlp.layers import TransformerEncoder
+
+
+def _roberta_kernel_initializer(stddev=0.02):
+    return keras.initializers.TruncatedNormal(stddev=stddev)
+
+
+class Roberta(keras.Model):
+    """Robustly optimized Bert pretraining approach.
+
+    This network implements a bi-directional Transformer-based encoder as
+    described in ["RoBERTa: A Robustly Optimized BERT Pretraining Approach"](https://arxiv.org/abs/1907.11692).
+    It includes the embedding lookups and transformer layers, but not include
+    the masked language model network.
+
+    This class gives a fully configurable Roberta model with any number of
+    layers, heads, and embedding dimensions. For specific roberta architectures
+    defined in the paper, see for example `keras_nlp.models.RobertaBase`.
+
+    Args:
+        vocabulary_size: Int. The size of the token vocabulary.
+        num_layers: Int. The number of transformer layers.
+        num_heads: Int. The number of attention heads for each transformer.
+            The hidden size must be divisible by the number of attention heads.
+        hidden_dim: Int. The size of the transformer encoding layer.
+        intermediate_dim: Int. The output dimension of the first Dense layer in
+            a two-layer feedforward network for each transformer.
+        dropout: Float. Dropout probability for the Transformer encoder.
+        max_sequence_length: Int. The maximum sequence length that this encoder
```

Comment:
Do we use `max_sequence_length` anywhere else? It seems to be just the ` sequence_length` arg from `TokenAndPositionEmbedding` and should probably be named the same.

"If None, `max_sequence_length` uses the value from sequence length" is not readily understandable to me. What is the expected behavior?

Reply:
Good catch! This comment was copied from the `Bert` class and I don't think it is correct there either. `max_sequence_length` must be specified as it controls the number of weights for queries, keys, and values. It cannot be inferred from the input and the function will error out if you pass `None`:
```
[/usr/local/lib/python3.7/dist-packages/keras_nlp/layers/position_embedding.py](https://localhost:8080/#) in __init__(self, sequence_length, initializer, **kwargs)
     77         if sequence_length is None:
     78             raise ValueError(
---> 79                 "`sequence_length` must be an Integer, received `None`."
     80             )
     81         self.sequence_length = int(sequence_length)

ValueError: `sequence_length` must be an Integer, received `None`.
```

Reply:
Now that we're on the subject, once we initialize the encoder with `max_sequence_length=12` (for example), you cannot pass inputs of a different size. I think this means that the shape of the input tensors should be `(max_sequence_length,)` instead of `(None,)`, right?
```
TypeError: Eager execution of tf.constant with unsupported shape. Tensor [0 0 0 0 0 1 1 1 1 1 0 0] (converted from [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0]) has 12 elements, but got `shape` (1, 11) with 11 elements).
```

Reply:
I think having the `None` shape is important? As I understand it the `max_sequence_length` is actually the length of the position embedding (it's the max number of positions that we can encode), which is 512 for all the BERT models. Probably similar or the same for RoBERTa?
https://github.com/google-research/bert/blob/master/modeling.py#L43

But it is actually import to allow people to pass in a different sequent length. Hence `shape=(None,)` on our inputs. Take a look at
https://www.tensorflow.org/tfmodels/nlp/fine_tune_bert#get_the_dataset_from_tensorflow_datasets
which uses 128 for fine tuning. This is an important point of flexibility, because using a 512 length will actually OOM the default colab GPU runtime.

Reply:
Hmm, I don't think our current implementation works unless the sequence length matches `max_sequence_length`. I'm pretty sure it's the job of the packer to pad the sequence to this length, right? If you want to pass a shorter sequence you have to use the attention mask to keep the model from processing it.

If I'm missing something big here that's ok, but I'm not sure our current implementation is right. You cannot pass a sequence of a different length without getting an exception.

Reply:
I believe this work on tensorflow/models impl (I haven't checked on ours), you can try out the colab for the guide I linked above. They do not pad to the max_sequence_length of the position embedding, the real inputs to the model are 128 long.

Note that the only variable in the entire model that uses the max sequence length is the position embedding, and that layer supports truncating the embedding to a shorter sequence length. So this should be able to work, and is likely something we need to support.
https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/position_embedding.py#L125

Reply:
I've played around with this a bit and this line will definitely error out unless every input is the same length:
```python
        x = keras.layers.Add()(
            (token_embedding, position_embedding, segment_embedding)
        )
```

So again let me know if our implementation is wrong...

Reply:
OK, looks like we commented at the same time. Should we punt for a followup PR to clean this up since it affects other files as well?

Reply:
I think this works today! With our BERT. Take a look at this colab
https://colab.sandbox.google.com/gist/mattdangerw/a3ee16c05c2e93cfa0055194026a6678/variable-length-inputs-to-a-bert-model.ipynb#scrollTo=vGSC6ii7Xvmy

All entries in a batch must be the same length. But we could actually vary the sequence length batch to batch if we wanted (not saying that's a good idea).

So conceptually we have a `max_sequence_length` that needs to agree with the length of the position embedding on the checkpoint we are loading, but does not need to agree with the sequence length of the actual input. I think what we have here is fine, though we could think about how to make this clearer in the docstring perhaps.

Reply:
Thanks for the clarification! Seems like this is a no-op, except for possibly improving the docstring.

Reply:
Yep! Side note, the reason this (variable sequence length inputs) is important is literally just memory usage.

Cutting the sequence length is an important trick to keep things running in the default colab GPU and other smaller accelerators. That's actually high priority for us, as colab will be the first touch for many of our lib.